File: app/src/main/java/dev/ragnarok/fenrir/colorpicker/slider/LightnessSlider.java
package dev.ragnarok.fenrir.colorpicker.slider;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.util.AttributeSet;

import dev.ragnarok.fenrir.colorpicker.ColorPickerView;
import dev.ragnarok.fenrir.colorpicker.Utils;
import dev.ragnarok.fenrir.colorpicker.builder.PaintBuilder;

public class LightnessSlider extends AbsCustomSlider {
    private final Paint barPaint = PaintBuilder.newPaint().build();
    private final Paint solid = PaintBuilder.newPaint().build();
    private final Paint clearingStroke = PaintBuilder.newPaint().color(0xffffffff).xPerMode(PorterDuff.Mode.CLEAR).build();
    private int color;
    private ColorPickerView colorPicker;

    public LightnessSlider(Context context) {
        super(context);
    }

    public LightnessSlider(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public LightnessSlider(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void drawBar(Canvas barCanvas) {
        int width = barCanvas.getWidth();
        int height = barCanvas.getHeight();

        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        int l = Math.max(2, width / 256);
        for (int x = 0; x <= width; x += l) {
            hsv[2] = (float) x / (width - 1);
            barPaint.setColor(Color.HSVToColor(hsv));
            barCanvas.drawRect(x, 0, x + l, height, barPaint);
        }
    }

    @Override
    protected void onValueChanged(float value) {
        if (colorPicker != null)
            colorPicker.setLightness(value);
    }

    @Override
    protected void drawHandle(Canvas canvas, float x, float y) {
        solid.setColor(Utils.colorAtLightness(color, value));
        if (showBorder) canvas.drawCircle(x, y, handleRadius, clearingStroke);
        canvas.drawCircle(x, y, handleRadius * 0.75f, solid);
    }

    public void setColorPicker(ColorPickerView colorPicker) {
        this.colorPicker = colorPicker;
    }

    public void setColor(int color) {
        this.color = color;
        value = Utils.lightnessOfColor(color);
        if (bar != null) {
            updateBar();
            invalidate();
        }
    }
}

File: preference/src/main/kotlin/de/maxr1998/modernpreferences/preferences/colorpicker/slider/LightnessSlider.kt
package de.maxr1998.modernpreferences.preferences.colorpicker.slider

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.PorterDuff
import android.util.AttributeSet
import de.maxr1998.modernpreferences.preferences.colorpicker.ColorPickerView
import de.maxr1998.modernpreferences.preferences.colorpicker.Utils
import de.maxr1998.modernpreferences.preferences.colorpicker.builder.PaintBuilder

class LightnessSlider : AbsCustomSlider {
    private val barPaint = PaintBuilder.newPaint().build()
    private val solid = PaintBuilder.newPaint().build()
    private val clearingStroke =
        PaintBuilder.newPaint().color(-0x1).xPerMode(PorterDuff.Mode.CLEAR).build()
    private var pColor = 0
    private var colorPicker: ColorPickerView? = null

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    override fun drawBar(barCanvas: Canvas?) {
        val width = (barCanvas ?: return).width
        val height = barCanvas.height
        val hsv = FloatArray(3)
        Color.colorToHSV(pColor, hsv)
        val l = 2.coerceAtLeast(width / 256)
        var x = 0
        while (x <= width) {
            hsv[2] = x.toFloat() / (width - 1)
            barPaint.color = Color.HSVToColor(hsv)
            barCanvas.drawRect(x.toFloat(), 0f, (x + l).toFloat(), height.toFloat(), barPaint)
            x += l
        }
    }

    override fun onValueChanged(value: Float) {
        colorPicker?.setLightness(value)
    }

    override fun drawHandle(canvas: Canvas, x: Float, y: Float) {
        solid.color =
            Utils.colorAtLightness(pColor, value)
        if (mShowBorder) canvas.drawCircle(x, y, handleRadius.toFloat(), clearingStroke)
        canvas.drawCircle(x, y, handleRadius * 0.75f, solid)
    }

    fun setColorPicker(colorPicker: ColorPickerView?) {
        this.colorPicker = colorPicker
    }

    fun setColor(pColor: Int) {
        this.pColor = pColor
        value = Utils.lightnessOfColor(pColor)
        if (bar != null) {
            updateBar()
            invalidate()
        }
    }
}